### PR TITLE
add nft recipts on deposit & withdraw

### PR DESCRIPTION
**Pull request by tac0turtle**

### CODE REVIEW

File: program/src/deposit.rs [hunk 1]
```diff
@@ -87,30 +96,50 @@ fn access_control(req: AccessControlRequest) -> Result<(), FundError> {
...
     vault_acc_info,
+    vault_authority_acc_info,
     depositor_authority_acc_info,
     token_program_acc_info,
+    nft_mint_acc_info,
+    nft_token_acc_info,
     amount,
   } = req;
 
+  if fund_acc.fund_type.eq(&FundType::PublicRaise) {
```
Comment (tac0turtle) [inline program/src/deposit.rs]:
```suggestion
{
    if fund_acc.fund_type.eq(&FundType::PublicRaise) {
```

File: src/accounts/fund.rs [hunk 2]
```diff
@@ -32,6 +32,15 @@ pub struct Fund {
   pub mint: Pubkey,
   /// Address of the token vault controlled by the Safe.
   pub vault: Pubkey,
+
+  /// Params
+
+  /// shares
+  pub shares: u64,
```
Comment (tac0turtle) [inline src/accounts/fund.rs:39]: this value should go up when funding goes up but not go down.

File: program/src/deposit.rs [hunk 3]
```diff
@@ -73,44 +82,71 @@ fn access_control(req: AccessControlRequest) -> Result<(), FundError> {
...
 
-  // let rent = access_control::rent(rent_acc_info)?;
-  let fund = access_control::fund(fund_acc_info, program_id)?;
-  let _ = access_control::vault_join(
-    vault_acc_info,
-    vault_authority_acc_info,
-    fund_acc_info,
-    program_id,
-  )?;
+    let _ = access_control::check_balance(fund_acc_info, amount)?;
```
Comment (tac0turtle) [inline program/src/deposit.rs:94]: should check if the fund is still open for deposits

File: program/src/withdraw.rs [hunk 4]
```diff
@@ -58,66 +65,89 @@ pub fn handler(
...
-  info!("invoking token transfer");
-  let withdraw_instruction = transfer(
-    &ID,
-    vault_acc_info.key,
-    withdraw_acc_info.key,
-    &vault_authority_acc_info.key,
-    &[],
-    amount as u64,
-  )?;
+  {
```
Comment (tac0turtle) [inline program/src/withdraw.rs:99]: when withdrawal happens it should be assumed the fund is not open anymore. There should be another handler for this specific tx.

File: program/src/deposit.rs (hunk 3 above)
Comment (cmwaters) [reply program/src/deposit.rs:94]: can a fund choose to stop accepting deposits?

Comment (tac0turtle) [reply program/src/deposit.rs:94]: for now on the first withdrawal it closes the deposits. Adding another tx to explicitly close funds.

File: src/accounts/fund.rs [hunk 5]
```diff
@@ -2,11 +2,11 @@ use serde::{Deserialize, Serialize};
 use serum_common::pack::*;
 use solana_client_gen::solana_sdk::pubkey::Pubkey;
 
-#[derive(Debug, Clone, Copy, Serialize, Deserialize)]
+#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq)]
 pub enum FundType {
   /// similar to a gofundme
   FundMe,
-  // Raise,
+  PublicRaise,
```
Comment (cmwaters) [inline src/accounts/fund.rs:9]: Why do you have this in two areas? One here and the other in src/types

Comment (tac0turtle) [reply src/accounts/fund.rs:9]: types.rs was moved here.

**Pull request merged**